packages/runed/src/lib/utilities/MediaQuery/MediaQuery.svelte.ts

```diff
@@ -34,43 +35,38 @@ import { browser } from "$lib/internal/utils/browser.js";
  * ```
  */
 export class MediaQuery {
-	#matches: boolean | undefined = $state();
+	#propQuery: MaybeGetter<string>;
+	#query = $derived.by(() => extract(this.#propQuery));
+	#mediaQueryList: MediaQueryList = $derived(window.matchMedia(this.#query));
 	#effectRegistered = false;
-	#query_fn: Getter<string | void> = () => {};
-	// this will be initialized in the constructor
-	#query = $derived(this.#query_fn?.()) as string;
+	#matches: boolean | undefined = $state();
 
 	constructor(query: MaybeGetter<string>) {
-		this.#query_fn = typeof query === "function" ? query : () => query;
-	}
-
-	#matchMedia() {
-		const result = window.matchMedia(this.#query);
-		this.#matches = result.matches;
-		return result;
+		this.#propQuery = query;
 	}
 
 	get matches(): boolean | undefined {
-		// if we are in an effect and this effect has not been registered yet
-		// we match the current value, register the listener and return match
 		if ($effect.active() && !this.#effectRegistered) {
+			this.#matches = this.#mediaQueryList.matches;
```

Comment:
you can avoid matching each time by using `$effect.pre`, which runs immediately for the first time, unlike `$effect`.
```ts
$effect.pre(() => {
  this.#matches = this.#mediaQueryList.matches;
  // rest of the effect
});
```